jetty-maven/jetty-maven-plugin/src/main/java/org/eclipse/jetty/maven/plugin/AbstractWebAppMojo.java

```diff
@@ -0,0 +1,463 @@
+import org.apache.maven.plugin.descriptor.PluginDescriptor;
+import org.apache.maven.plugins.annotations.Component;
+import org.apache.maven.plugins.annotations.Parameter;
+import org.apache.maven.project.MavenProject;
+import org.codehaus.plexus.classworlds.realm.ClassRealm;
+import org.codehaus.plexus.configuration.PlexusConfiguration;
+import org.codehaus.plexus.util.xml.Xpp3Dom;
+import org.eclipse.aether.RepositorySystem;
+import org.eclipse.aether.RepositorySystemSession;
+import org.eclipse.aether.repository.RemoteRepository;
+import org.eclipse.jetty.maven.MavenServerConnector;
+import org.eclipse.jetty.maven.ScanTargetPattern;
+import org.eclipse.jetty.maven.utils.MavenProjectHelper;
+import org.eclipse.jetty.server.RequestLog;
+import org.eclipse.jetty.server.Server;
+import org.eclipse.jetty.server.handler.ContextHandler;
+import org.eclipse.jetty.util.resource.Resource;
+import org.twdata.maven.mojoexecutor.MojoExecutor;
+
+/**
+ * AbstractWebAppMojo
+ *
+ * Base class for common behaviour of jetty mojos.
+ */
+public abstract class AbstractWebAppMojo extends AbstractMojo
+{
+
+    @Inject
+    protected BuildPluginManager buildPluginManager;
+
+    public enum DeploymentMode 
+    {
+        EMBED,
+        FORK,
+        HOME, //alias for EXTERNAL
+        DISTRO, //alias for EXTERNAL
+        EXTERNAL
+    }
+
+    /**
+     * jetty environment value such ee8, ee9, ee10
+     * if empty will be discovered from your dependencies
+     */
+    @Parameter (defaultValue = "ee10", property = "jetty.environment")
+    private String environment;
+
+    /**
+     * Max number of times to check to see if jetty has started correctly
+     * when running in FORK or EXTERNAL mode.
+     */
+    @Parameter (defaultValue = "10")
+    protected int maxChildStartChecks;
+
+    /**
+     * How long to wait in msec between checks to see if jetty has started
+     * correctly when running in FORK or EXTERNAL mode.
+     */
+    @Parameter (defaultValue = "200")
+    protected long maxChildStartCheckMs;
+
+    /**
+     * Whether or not to include dependencies on the plugin's classpath with &lt;scope&gt;provided&lt;/scope&gt;
+     * Use WITH CAUTION as you may wind up with duplicate jars/classes.
+     * 
+     * @since jetty-7.5.2
+     */
+    @Parameter (defaultValue = "false")
+    protected boolean useProvidedScope;
+
+    /**
+     * List of goals that are NOT to be used
+     * 
+     * @since jetty-7.5.2
+     */
+    @Parameter
+    protected String[] excludedGoals;
+
+    /**
+     * An instance of org.eclipse.jetty.eeXX.webapp.WebAppContext that represents the webapp.
+     * Use any of its setters to configure the webapp. This is the preferred and most
+     * flexible method of configuration, rather than using the (deprecated) individual
+     * parameters like "tmpDirectory", "contextPath" etc.
+     * 
+     */
+    @Parameter(name = "webApp")
+    protected PlexusConfiguration webApp;
```

Comment:
The comment is wrong here, it's not a `WebAppContext`.

Is the idea that you can use this `PlexusConfiguration` to hold any old `name,value` pairs, and that during the invocation of the target plugin, any `name,value` pairs that match those of the setters on the target `WebAppContext` will be transferred?

 In that case my reservation is that there's no type checking on what the user is configuring in the pom.xml. If we change eg `ee10` `WebAppContext` then the user won't get an error.

Reply:
> The comment is wrong here, it's not a `WebAppContext`.
> 
> Is the idea that you can use this `PlexusConfiguration` to hold any old `name,value` pairs, and that during the invocation of the target plugin, any `name,value` pairs that match those of the setters on the target `WebAppContext` will be transferred?
> 
> In that case my reservation is that there's no type checking on what the user is configuring in the pom.xml. If we change eg `ee10` `WebAppContext` then the user won't get an error.

maven will still check when running the target mojo.

but yes the problem is this inheritance of `org.eclipse.jetty.ee9.maven.plugin.MavenWebAppContext extends WebAppContext` (same for ee10)
because classes which are mojo fields should not inherit but should be only bean carrying some configuration which would be injected when running the target eeX. 
So this would simplify a lot of code which is already duplicated (not only here) and simplify the code.. 
With exposing those classes we create some problems to ourself as every changes in the parent class `WebAppContext` can potentially create issues for users. we should hide that and simplify the configuration (again users wants some very specific they can still use a `jetty-**.xml` file.